Build issue with HPX_WITH_NETWORKING=OFF

The current HPX master branch ( 1cbb0d6e15dcb6a8bf841250d235f436fbe6a432 ) does not compile when HPX_WITH_NETWORKING is set to OFF (using clang 12, single node CUDA build). This appears to be a new issue, as the same build without networking works using HPX 1.9.0.


Setting HPX_WITH_NETWORKING=ON fixes the compilation on master as well, with OFF I get the following compilation errors:


```src/hpx/libs/full/runtime_distributed/src/runtime_distributed.cpp:453:17: error: member access into incomplete type 'hpx::util::io_service_pool'
        io_pool_->run(false);
                ^
src/hpx/libs/core/io_service/include/hpx/io_service/io_service_pool_fwd.hpp:13:27: note: forward declaration of 'hpx::util::io_service_pool'
    class HPX_CORE_EXPORT io_service_pool;
                          ^
src/hpx/libs/full/runtime_distributed/src/runtime_distributed.cpp:555:19: error: member access into incomplete type 'hpx::util::io_service_pool'
        main_pool_->stop();
                  ^
src/hpx/libs/core/io_service/include/hpx/io_service/io_service_pool_fwd.hpp:13:27: note: forward declaration of 'hpx::util::io_service_pool'
    class HPX_CORE_EXPORT io_service_pool;
                          ^
src/hpx/libs/full/runtime_distributed/src/runtime_distributed.cpp:579:19: error: member access into incomplete type 'hpx::util::io_service_pool'
        main_pool_->thread_run(0);
                  ^
src/hpx/libs/core/io_service/include/hpx/io_service/io_service_pool_fwd.hpp:13:27: note: forward declaration of 'hpx::util::io_service_pool'
    class HPX_CORE_EXPORT io_service_pool;
                          ^
src/hpx/libs/full/runtime_distributed/src/runtime_distributed.cpp:643:20: error: member access into incomplete type 'hpx::util::io_service_pool'
        timer_pool_->stop();
                   ^
src/hpx/libs/core/io_service/include/hpx/io_service/io_service_pool_fwd.hpp:13:27: note: forward declaration of 'hpx::util::io_service_pool'
    class HPX_CORE_EXPORT io_service_pool;
                          ^
src/hpx/libs/full/runtime_distributed/src/runtime_distributed.cpp:646:24: error: member access into incomplete type 'hpx::util::io_service_pool'
            timer_pool_->join();
                       ^
src/hpx/libs/core/io_service/include/hpx/io_service/io_service_pool_fwd.hpp:13:27: note: forward declaration of 'hpx::util::io_service_pool'
    class HPX_CORE_EXPORT io_service_pool;
                          ^
src/hpx/libs/full/runtime_distributed/src/runtime_distributed.cpp:647:24: error: member access into incomplete type 'hpx::util::io_service_pool'
            timer_pool_->clear();
                       ^
src/hpx/libs/core/io_service/include/hpx/io_service/io_service_pool_fwd.hpp:13:27: note: forward declaration of 'hpx::util::io_service_pool'
    class HPX_CORE_EXPORT io_service_pool;
                          ^
src/hpx/libs/full/runtime_distributed/src/runtime_distributed.cpp:652:17: error: member access into incomplete type 'hpx::util::io_service_pool'
        io_pool_->stop();
                ^
src/hpx/libs/core/io_service/include/hpx/io_service/io_service_pool_fwd.hpp:13:27: note: forward declaration of 'hpx::util::io_service_pool'
    class HPX_CORE_EXPORT io_service_pool;
                          ^
src/hpx/libs/full/runtime_distributed/src/runtime_distributed.cpp:655:21: error: member access into incomplete type 'hpx::util::io_service_pool'
            io_pool_->join();
                    ^
src/hpx/libs/core/io_service/include/hpx/io_service/io_service_pool_fwd.hpp:13:27: note: forward declaration of 'hpx::util::io_service_pool'
    class HPX_CORE_EXPORT io_service_pool;
                          ^
src/hpx/libs/full/runtime_distributed/src/runtime_distributed.cpp:656:21: error: member access into incomplete type 'hpx::util::io_service_pool'
            io_pool_->clear();
                    ^
src/hpx/libs/core/io_service/include/hpx/io_service/io_service_pool_fwd.hpp:13:27: note: forward declaration of 'hpx::util::io_service_pool'
    class HPX_CORE_EXPORT io_service_pool;
                          ^
9 errors generated.
```

--- a/.jenkins/lsu/env-gcc-11.sh
+++ b/.jenkins/lsu/env-gcc-11.sh
@@ -20,9 +20,7 @@ configure_extra_options+=" -DHPX_WITH_MALLOC=system"
 configure_extra_options+=" -DHPX_WITH_FETCH_ASIO=ON"
 configure_extra_options+=" -DHPX_WITH_COMPILER_WARNINGS=ON"
 configure_extra_options+=" -DHPX_WITH_COMPILER_WARNINGS_AS_ERRORS=ON"
-configure_extra_options+=" -DHPX_WITH_PARCELPORT_MPI=ON"
-configure_extra_options+=" -DHPX_WITH_PARCELPORT_LCI=ON"
-configure_extra_options+=" -DHPX_WITH_FETCH_LCI=ON"
+configure_extra_options+=" -DHPX_WITH_NETWORKING=OFF"
 configure_extra_options+=" -DHPX_WITH_DATAPAR_BACKEND=STD_EXPERIMENTAL_SIMD"
 
 # The pwrapi library still needs to be set up properly on rostam

--- a/examples/quickstart/custom_serialization.cpp
+++ b/examples/quickstart/custom_serialization.cpp
@@ -203,8 +203,16 @@ int main()
 
     send_gravity_action gravityAction;
     auto gravity = planet_weight_calculator(9.81);
-    hpx::async(gravityAction, hpx::find_remote_localities()[0], gravity).get();
 
+    auto remote_localities = hpx::find_remote_localities();
+    if (!remote_localities.empty())
+    {
+        hpx::async(gravityAction, remote_localities[0], gravity).get();
+    }
+    else
+    {
+        hpx::async(gravityAction, hpx::find_here(), gravity).get();
+    }
     return 0;
 }
 //]

--- a/libs/full/agas/src/addressing_service.cpp
+++ b/libs/full/agas/src/addressing_service.cpp
@@ -736,8 +736,10 @@ namespace hpx::agas {
         naming::gid_type const id(
             naming::detail::get_stripped_gid_except_dont_cache(gid));
 
+#if defined(HPX_HAVE_NETWORKING)
         // migratable objects should be handled by the function below
         HPX_ASSERT(!naming::detail::is_migratable(gid));
+#endif
 
         // Try to resolve the address of the GID from the locally available
         // information.

--- a/libs/full/runtime_distributed/include/hpx/runtime_distributed/migrate_component.hpp
+++ b/libs/full/runtime_distributed/include/hpx/runtime_distributed/migrate_component.hpp
@@ -49,16 +49,21 @@ namespace hpx { namespace components {
 #if defined(DOXYGEN)
     future<hpx::id_type>
 #else
-    inline std::enable_if_t<traits::is_component<Component>::value &&
-            traits::is_distribution_policy<DistPolicy>::value,
+    std::enable_if_t<traits::is_component_v<Component> &&
+            traits::is_distribution_policy_v<DistPolicy>,
         future<hpx::id_type>>
 #endif
-    migrate(hpx::id_type const& to_migrate, DistPolicy const& policy)
+    migrate(hpx::id_type const& to_migrate,
+        [[maybe_unused]] DistPolicy const& policy)
     {
+#if defined(HPX_HAVE_NETWORKING)
         using action_type =
             server::perform_migrate_component_action<Component, DistPolicy>;
         return hpx::detail::async_colocated<action_type>(
             to_migrate, to_migrate, policy);
+#else
+        return hpx::make_ready_future(to_migrate);
+#endif
     }
 
     /// Migrate the given component to the specified target locality
@@ -133,8 +138,7 @@ namespace hpx { namespace components {
 #if defined(DOXYGEN)
     future<hpx::id_type>
 #else
-    inline std::enable_if_t<traits::is_component<Component>::value,
-        future<hpx::id_type>>
+    std::enable_if_t<traits::is_component_v<Component>, future<hpx::id_type>>
 #endif
     migrate(hpx::id_type const& to_migrate, hpx::id_type const& target_locality)
     {

--- a/libs/full/runtime_distributed/src/runtime_distributed.cpp
+++ b/libs/full/runtime_distributed/src/runtime_distributed.cpp
@@ -23,6 +23,7 @@
 #include <hpx/functional/function.hpp>
 #include <hpx/itt_notify/thread_name.hpp>
 #include <hpx/modules/errors.hpp>
+#include <hpx/modules/io_service.hpp>
 #include <hpx/modules/logging.hpp>
 #include <hpx/modules/static_reinit.hpp>
 #include <hpx/modules/threadmanager.hpp>

